model_server/src/model_server/app.py

```diff
@@ -0,0 +1,49 @@
+import asyncio
+from json import JSONDecodeError
+from starlette.applications import Starlette
+from starlette.responses import JSONResponse
+from starlette.routing import Route
+from transformers import pipeline
+
+from model_server import settings
+from model_server.handler import server_loop
+
+
+def handler(command):
+    """Make a command specific handler."""
+    async def handle(request):
+        try:
+            payload = await request.json()
+        except JSONDecodeError:
+            return JSONResponse({'error': 'invalid JSON provided'}, status_code=400)
+
+        # The model should only be loaded into memory once, seeing it's large.
+        # Use a queue to send/receive messages from it.
+        response_q = asyncio.Queue()
+        await request.app.model_queue.put((command, payload, response_q))
+        output, status = await response_q.get()
+        return JSONResponse(output, status_code=status)
+
+    return handle
+
+
+routes = [
+    Route("/", handler(settings.DEFAULT), methods=["POST"]),
+    Route("/encoding", handler(settings.ENCODE), methods=["POST"]),
+    Route("/question_answering", handler(settings.QUESTION_ANSWERING), methods=["POST"]),
+    Route("/paraphrase_mining", handler(settings.PARAPHRASE_MINING), methods=["POST"]),
+]
```

Comment:
It looks like all models expose 4 endpoints: default, encoding, question_answering and paraphrase_mining? It feels a bit more complicated than necessary but maybe it's clearer to have explicit endpoints rather than only defaults? Since it's for our internal use that's probably ok. Just to be clear:
- Encoders (retriever & lit) by default encode but can also do paraphrase_mining. 
- qa_model ONLY supports question_answering.

> That does make things a lot clearer :D I changed the reader-model to qa-model and added a question_answering endpoint for encoders

Encoders don't support question_answering but rather paraphrase_mining endpoint

Reply:
hmm. An alternative would be to have the routes list get created on the basis of the model type? Or have a sort of plugin type thingy, where each model type has its own module (or class) and depending on what MODEL_TYPE is set, the appropriate module gets used... 

Yup, paraphrase_mining. I'll get the names right someday :/

Reply:
I think customizing by MODEL_TYPE would be better. The encoder(s) and qa_models are actually fundamentally quite different. Exposing the same endpoints doesn't "feel" right.

And I totally get the feeling. After a while, everything sounds like blah blah blah...

Reply:
Totally overhauled the code - now there is a per model type module which will encapsulate the whole functionality of that model

Reply:
Great!